### ISSUE TITLE
[V14.0.20211227] Permission error on reports with nested template

### ISSUE THREAD

**Issue by alessandro-fiorino** [odoo/odoo#81960]
Impacted versions:
V14.0.20211227, V14.0.20211220 was ok

Steps to reproduce:
Create a report with nested template calls and generate it from a non-admin account

Current behavior:
A traceback like the one at the end of the message

Expected behavior:
It worked perfectly previously

Support ticket number submitted via odoo.com/help (optional): 2723190

I think the culprit is this code snippet from ir_actions_report line 859 that changed between the two releases:
```
        if report_model is not None:
            # _render_ may be executed in sudo but evaluation context as real user
            report_model = report_model.sudo(False)
            data.update(report_model._get_report_values(docids, data=data))
        else:
            # _render_ may be executed in sudo but evaluation context as real user
            docs = self.env[self.model].sudo(False).browse(docids)
            data.update({
                'doc_ids': docids,
                'doc_model': self.model,
                'docs': docs,
            })
        return data
```
before it was
```
        if report_model is not None:
            data.update(report_model._get_report_values(docids, data=data))
        else:
            docs = self.env[self_sudo.model].browse(docids)
            data.update({
                'doc_ids': docids,
                'doc_model': self_sudo.model,
                'docs': docs,
            })
        return data
```
The traceback is
```
Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/api.py", line 793, in get
return field_cache[record._ids[0]]
KeyError: 213

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/fields.py", line 972, in __get__
value = env.cache.get(record, self)
File "/home/odoo/src/odoo/odoo/api.py", line 796, in get
raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'ir.ui.view(213,).key'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 85, in lookup
r = d[key]
File "/home/odoo/src/odoo/odoo/tools/func.py", line 71, in wrapper
return func(self, *args, **kwargs)
File "/home/odoo/src/odoo/odoo/tools/lru.py", line 34, in __getitem__
a = self.d[obj]
KeyError: ('ir.model.access', <function IrModelAccess.check at 0x7f0264e73c10>, 163, False, 'ir.ui.view', 'read', True, ('it_IT',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/fields.py", line 998, in __get__
recs._fetch_field(self)
File "/home/odoo/src/odoo/odoo/models.py", line 3071, in _fetch_field
self._read(fnames)
File "/home/odoo/src/odoo/odoo/models.py", line 3085, in _read
self.check_access_rights('read')
File "/home/odoo/src/odoo/odoo/models.py", line 3332, in check_access_rights
return self.env['ir.model.access'].check(self._name, operation, raise_exception)
File "<decorator-gen-34>", line 2, in check
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 90, in lookup
value = d[key] = self.method(*args, **kwargs)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1805, in check
raise AccessError(msg)
odoo.exceptions.AccessError: Accesso ai record "Vista" (ir.ui.view) non consentito.

L'operazione è consentita per i seguenti gruppi:
- Administration/Settings
- Website/Editor and Designer
- Website/Restricted Editor

Se necessario, contattare l'amministratore per richiedere l'accesso.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 85, in lookup
r = d[key]
File "/home/odoo/src/odoo/odoo/tools/func.py", line 71, in wrapper
return func(self, *args, **kwargs)
File "/home/odoo/src/odoo/odoo/tools/lru.py", line 34, in __getitem__
a = self.d[obj]
KeyError: ('ir.model.access', <function IrModelAccess.check at 0x7f0264e73c10>, 163, False, 'ir.ui.view', 'read', True, ('it_IT',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 331, in _compiled_fn
return compiled(self, append, new, options, log)
File "<template>", line 1, in template_web_external_layout_382
File "/home/odoo/src/odoo/odoo/fields.py", line 1000, in __get__
record._fetch_field(self)
File "/home/odoo/src/odoo/odoo/models.py", line 3071, in _fetch_field
self._read(fnames)
File "/home/odoo/src/odoo/odoo/models.py", line 3085, in _read
self.check_access_rights('read')
File "/home/odoo/src/odoo/odoo/models.py", line 3332, in check_access_rights
return self.env['ir.model.access'].check(self._name, operation, raise_exception)
File "<decorator-gen-34>", line 2, in check
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 90, in lookup
value = d[key] = self.method(*args, **kwargs)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1805, in check
raise AccessError(msg)
odoo.exceptions.AccessError: Accesso ai record "Vista" (ir.ui.view) non consentito.

L'operazione è consentita per i seguenti gruppi:
- Administration/Settings
- Website/Editor and Designer
- Website/Restricted Editor

Se necessario, contattare l'amministratore per richiedere l'accesso.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/addons/web/controllers/main.py", line 2124, in report_download
response = self.report_routes(reportname, docids=docids, converter=converter, context=context)
File "/home/odoo/src/odoo/odoo/http.py", line 532, in response_wrap
response = f(*args, **kw)
File "/home/odoo/src/odoo/addons/web/controllers/main.py", line 2059, in report_routes
pdf = report.with_context(context)._render_qweb_pdf(docids, data=data)[0]
File "/home/odoo/src/odoo/addons/account/models/ir_actions_report.py", line 50, in _render_qweb_pdf
return super()._render_qweb_pdf(res_ids=res_ids, data=data)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_actions_report.py", line 805, in _render_qweb_pdf
html = self_sudo.with_context(context)._render_qweb_html(res_ids, data=data)[0]
File "/home/odoo/src/enterprise/web_studio/models/ir_actions_report.py", line 18, in _render_qweb_html
return super(IrActionsReport, self)._render_qweb_html(docids, data)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_actions_report.py", line 846, in _render_qweb_html
return self._render_template(self.report_name, data), 'html'
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_actions_report.py", line 582, in _render_template
return view_obj._render_template(template, values)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 1708, in _render_template
return self.browse(self.get_view_id(template))._render(values, engine)
File "/home/odoo/src/odoo/addons/website/models/ir_ui_view.py", line 420, in _render
return super(View, self)._render(values, engine=engine, minimal_qcontext=minimal_qcontext)
File "/home/odoo/src/odoo/addons/web_editor/models/ir_ui_view.py", line 28, in _render
return super(IrUiView, self)._render(values=values, engine=engine, minimal_qcontext=minimal_qcontext)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_ui_view.py", line 1716, in _render
return self.env[engine]._render(self.id, qcontext)
File "/home/odoo/src/enterprise/web_studio/models/ir_qweb.py", line 44, in _render
return super(IrQWeb, self)._render(template, values=values, **options)
File "/home/odoo/src/user/OCA/web/web_company_color/models/ir_qweb.py", line 79, in _render
return super()._render(id_or_xml_id, values=values, **options)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_qweb.py", line 55, in _render
result = super(IrQWeb, self)._render(id_or_xml_id, values=values, **context)
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 258, in _render
self.compile(template, options)(self, body.append, values or {})
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 333, in _compiled_fn
raise e
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 331, in _compiled_fn
return compiled(self, append, new, options, log)
File "<template>", line 1, in template_waste_report_analysis_lot_342
File "<template>", line 2, in body_call_content_340
File "<template>", line 3, in foreach_339
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 333, in _compiled_fn
raise e
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 331, in _compiled_fn
return compiled(self, append, new, options, log)
File "<template>", line 1, in template_waste_report_analysis_lot_document_371
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 338, in _compiled_fn
raise QWebException("Error to render compiling AST", e, path, node and etree.tostring(node[0], encoding='unicode'), name)
odoo.addons.base.models.qweb.QWebException: Accesso ai record "Vista" (ir.ui.view) non consentito.

L'operazione è consentita per i seguenti gruppi:
- Administration/Settings
- Website/Editor and Designer
- Website/Restricted Editor

Se necessario, contattare l'amministratore per richiedere l'accesso.
Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/api.py", line 793, in get
return field_cache[record._ids[0]]
KeyError: 213

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/fields.py", line 972, in __get__
value = env.cache.get(record, self)
File "/home/odoo/src/odoo/odoo/api.py", line 796, in get
raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'ir.ui.view(213,).key'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 85, in lookup
r = d[key]
File "/home/odoo/src/odoo/odoo/tools/func.py", line 71, in wrapper
return func(self, *args, **kwargs)
File "/home/odoo/src/odoo/odoo/tools/lru.py", line 34, in __getitem__
a = self.d[obj]
KeyError: ('ir.model.access', <function IrModelAccess.check at 0x7f0264e73c10>, 163, False, 'ir.ui.view', 'read', True, ('it_IT',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/fields.py", line 998, in __get__
recs._fetch_field(self)
File "/home/odoo/src/odoo/odoo/models.py", line 3071, in _fetch_field
self._read(fnames)
File "/home/odoo/src/odoo/odoo/models.py", line 3085, in _read
self.check_access_rights('read')
File "/home/odoo/src/odoo/odoo/models.py", line 3332, in check_access_rights
return self.env['ir.model.access'].check(self._name, operation, raise_exception)
File "<decorator-gen-34>", line 2, in check
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 90, in lookup
value = d[key] = self.method(*args, **kwargs)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1805, in check
raise AccessError(msg)
odoo.exceptions.AccessError: Accesso ai record "Vista" (ir.ui.view) non consentito.

L'operazione è consentita per i seguenti gruppi:
- Administration/Settings
- Website/Editor and Designer
- Website/Restricted Editor

Se necessario, contattare l'amministratore per richiedere l'accesso.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 85, in lookup
r = d[key]
File "/home/odoo/src/odoo/odoo/tools/func.py", line 71, in wrapper
return func(self, *args, **kwargs)
File "/home/odoo/src/odoo/odoo/tools/lru.py", line 34, in __getitem__
a = self.d[obj]
KeyError: ('ir.model.access', <function IrModelAccess.check at 0x7f0264e73c10>, 163, False, 'ir.ui.view', 'read', True, ('it_IT',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/odoo/src/odoo/odoo/addons/base/models/qweb.py", line 331, in _compiled_fn
return compiled(self, append, new, options, log)
File "<template>", line 1, in template_web_external_layout_382
File "/home/odoo/src/odoo/odoo/fields.py", line 1000, in __get__
record._fetch_field(self)
File "/home/odoo/src/odoo/odoo/models.py", line 3071, in _fetch_field
self._read(fnames)
File "/home/odoo/src/odoo/odoo/models.py", line 3085, in _read
self.check_access_rights('read')
File "/home/odoo/src/odoo/odoo/models.py", line 3332, in check_access_rights
return self.env['ir.model.access'].check(self._name, operation, raise_exception)
File "<decorator-gen-34>", line 2, in check
File "/home/odoo/src/odoo/odoo/tools/cache.py", line 90, in lookup
value = d[key] = self.method(*args, **kwargs)
File "/home/odoo/src/odoo/odoo/addons/base/models/ir_model.py", line 1805, in check
raise AccessError(msg)
odoo.exceptions.AccessError: Accesso ai record "Vista" (ir.ui.view) non consentito.

L'operazione è consentita per i seguenti gruppi:
- Administration/Settings
- Website/Editor and Designer
- Website/Restricted Editor

Se necessario, contattare l'amministratore per richiedere l'accesso.

Error to render compiling AST
AccessError: Accesso ai record "Vista" (ir.ui.view) non consentito.

L'operazione è consentita per i seguenti gruppi:
- Administration/Settings
- Website/Editor and Designer
- Website/Restricted Editor

Se necessario, contattare l'amministratore per richiedere l'accesso.
Template: web.external_layout
Path: /t/t[3]/t
Node: <t t-raw="0"/>
```

**Comment by hildickethan** [comment]
I have the same issue, it seems https://github.com/odoo/odoo/commit/f2c1ee5a622db33a4411e7f9285f09387d1d7480 is what causes it like you mentioned.
Backporting this 15.0 fix https://github.com/odoo/odoo/pull/80510 works to fix it, but the fact that both of these commits are old makes me think I'm missing something

**Comment by alessandro-fiorino** [comment]
On the odoo.sh managed version the patch introducing the bug has been appliend between 2021-12-20 and 2021-12-27, maybe the commit is old but on that platform it showed up only in the last days.

**Comment by mart-e** [comment]
Thanks for the report, I have backported the patch at #82050

**Comment by hildickethan** [comment]
Thanks for the patch

**Comment by odony** [comment]
Closing as it seems fixed by #82050 :)
